src/runner.rs

```diff
@@ -32,12 +34,36 @@ impl SubsystemRunner {
         Fut: 'static + Future<Output = Result<(), Err>> + Send,
         Err: Into<ErrType>,
     {
-        let future = async { run_subsystem(name, subsystem, subsystem_handle, guard).await };
-        let aborthandle = tokio::spawn(future).abort_handle();
+        let future = {
+            let name = name.clone();
+            async move { run_subsystem(name, subsystem, subsystem_handle, guard).await }
+        };
+
+        let aborthandle = spawn(future, name);
         SubsystemRunner { aborthandle }
     }
 }
 
+#[cfg(not(feature = "tokio-unstable"))]
+fn spawn<F: Future + Send + 'static>(f: F, _name: Arc<str>) -> AbortHandle
+where
+    <F as Future>::Output: Send,
+{
+    tokio::spawn(f).abort_handle()
+}
+
+#[cfg(feature = "tokio-unstable")]
+fn spawn<F: Future + Send + 'static>(f: F, name: Arc<str>) -> AbortHandle
+where
+    <F as Future>::Output: Send,
+{
+    tokio::task::Builder::new()
+        .name(&name)
+        .spawn(f)
+        .expect("spawning a task does not fail")
```

Comment:
If you are entirely sure it will never fail, you can use "unwrap". Otherwise, we should do some proper error handling. But as it is currently, it's quite confusing, because it will print "panicked: does not fail" ;)

Reply:
I thought there was a discussion about that in the rust community. The message should exactly show what was expected and why that expectation was broken.

That's also what the docs [recommend](https://doc.rust-lang.org/stable/std/option/enum.Option.html#recommended-message-style)

Btw. I don't know why `tokio::spawn` does return a plain JoinHandle and the Builder does return a Result. Also no hint for failure cases in the docs.

Reply:
OK, I agree that we should use `expect` then, but "does not fail" is not a reason why it should not fail ;) especially if you don't know the reason why they added an `Err` there

Reply:
It actually never returns Err, it panics if it's run outside a tokio-runtime.

https://github.com/tokio-rs/tokio/blob/master/tokio/src/task/builder.rs#L93

Reply:
I'm interested, so I asked.

https://github.com/tokio-rs/tokio/discussions/6946

Reply:
OK :) then change the text to something more meaningful and I'll accept it. Like "a task should be spawned". "does not fail" is not something you expect, it's a certainty, and is an anti pattern for a `Result`, which indicates the opposite of certainty. While it might not be possible now, I'm sure they added it for potential future use.